File: day2/day2.R
library(data.table)

### DAY 2

## PART 1

data <- fread("data.txt", header = F, col.names = c("opponent", "me"))

winning <- data.table(
  op = c("A", "B", "C"),
  me = c("Y", "Z", "X")
)

draw <- data.table(
  op = c("A", "B", "C"),
  me = c("X", "Y", "Z")
)

hand.score <- data.table(
  me = c("X", "Y", "Z"),
  score = c(1, 2, 3)
)

wins <- apply(data, 1, paste, collapse = "_") %in% apply(winning, 1, paste, collapse = "_")
draws <- apply(data, 1, paste, collapse = "_") %in% apply(draw, 1, paste, collapse = "_")

cat("If my strategy was correct, my total score would be ",
    sum(draws)*3 + sum(wins)*6 + sum(merge(hand.score, data)$score),
    ".\n",
    sep = "")


## PART 2

data <- fread("data.txt", header = F, col.names = c("opponent", "outcome"))

losing <- data.table(
  op = c("A", "B", "C"),
  me = c("Z", "X", "Y")
)

setkey(hand.score, "me")
setkey(winning, "op")
setkey(draw, "op")
setkey(losing, "op")

cat("If the elf's strategy was correct, my total score would be ",
    sum(c(hand.score[losing[data[outcome == "X", opponent]]$me, score], 
          hand.score[draw[data[outcome == "Y", opponent]]$me, score] + 3, 
          hand.score[winning[data[outcome == "Z", opponent]]$me, score] + 6)),
    ".\n",
    sep = "")

File: day2/day2.py
# -*- coding: utf-8 -*-

### DAY 2

## PART 1

def getScoreUsingSign(me, opponent):
    match me:
        case "X":
            winning = "C"
            draw = "A"
            losing = "B"
            score = 1
        case "Y":
            winning = "A"
            draw = "B"
            losing = "C"
            score = 2
        case "Z":
            winning = "B"
            draw = "C"
            losing = "A"
            score = 3

    if opponent == winning:
        score += 6
    elif opponent == draw:
        score += 3
    elif opponent == losing:
        score

    return(score)

with open(r"data.txt", "r") as f:
    scores = 0
    for line in f:
        opponent = line.split(" ")[0]
        me = line.split(" ")[1].split("\n")[0]

        scores += getScoreUsingSign(me, opponent)

print(f"If my strategy was correct, my total score would be {scores}.")


## PART 2

def getScoreUsingOutcome(outcome, opponent):
    match opponent:
        case "A":
            winning = "Y"
            draw = "X"
            losing = "Z"
        case "B":
            winning = "Z"
            draw = "Y"
            losing = "X"
        case "C":
            winning = "X"
            draw = "Z"
            losing = "Y"

    match outcome:
        case "X":
            score = getScoreUsingSign(losing, opponent)
        case "Y":
            score = getScoreUsingSign(draw, opponent)
        case "Z":
            score = getScoreUsingSign(winning, opponent)

    return(score)

with open(r"data.txt", "r") as f:
    scores = 0
    for line in f:
        opponent = line.split(" ")[0]
        outcome = line.split(" ")[1].split("\n")[0]

        scores += getScoreUsingOutcome(outcome, opponent)

print(f"If the elf's strategy was correct, my total score would be {scores}.")